trieste/acquisition/function.py

```diff
@@ -130,6 +133,117 @@ def expected_improvement(model: ModelInterface, eta: tf.Tensor, at: QueryPoints)
     return (eta - mean) * normal.cdf(eta) + variance * normal.prob(eta)
 
 
+class MaxValueEntropySearch(SingleModelAcquisitionBuilder):
+    """
+    Builder for the max-value entropy search acqusiiton function (for function minimisation)
+    """
+
+    def __init__(self,search_space: SearchSpace, num_samples: int = 10, grid_size: int = 5000):
+        """
+        :param num_samples: integer determining how many samples to draw of the minimum 
+            (does not need to be large).
+        :param grid_size: number of random locations in grid used to fit the gumbel distribution 
+            and approximately generate the samples of the minimum (recommend scaling with problem dimension).
+        """
+        self._num_samples = num_samples
+        self._grid_size = grid_size
+        self._search_space = search_space
+
+    def prepare_acquisition_function(
+        self, dataset: Dataset, model: ModelInterface
+    ) -> AcquisitionFunction:
+        """
+        Need to sample possible min-values from our posterior. To do this we implement a Gumbel sampler.
+        we approximate Pr(y*^hat<y) by Gumbel(alpha,beta) then sample from Gumbel.
+
+        :param dataset: Unused.
+        :param model: The model over the specified ``dataset``.
+        :return: The MES function.
+        """
+
+        # generate random grid to fit Gumbel distribution
+        # grid consists of already queried locations and grid_size other random locations
+        query_points = self._search_space.sample(self._grid_size)
+        query_points = tf.concat([dataset.query_points,query_points],0)
+        fmean, fvar = model.predict(query_points)
+        fsd = tf.math.sqrt(fvar)
+        idx = tf.math.argmax(-fmean[:tf.shape(dataset.query_points)[0]])
+
+        # Fit Gumbel distriubtion
+        # scaling so that gumbel scale is proportional to IQ range of cdf Pr(y*<z)
+        # find quantiles Pr(y*<y1)=r1 and Pr(y*<y2)=r2
+        right = tf.squeeze(fmean[tf.squeeze(idx)])
+        left = right
+        probf = lambda x: tf.math.exp(tf.math.reduce_sum(tfp.distributions.Normal(tf.cast(0,tf.float64),tf.cast(1,tf.float64)).log_cdf(-(x - fmean) / fsd), axis=0))
+        
+        # get range for binary search
+        i = 0
+        while probf(left) < 0.75:
+            left = 2. ** i * tf.math.reduce_min(fmean - 5. * fsd) + (1. - 2. ** i) * right
+            i += 1
+        i = 0
+        while probf(right) > 0.25:
+            right = -2. ** i * tf.math.reduce_min(fmean - 5. * fsd) + (1. + 2. ** i) * tf.squeeze(fmean[tf.squeeze(idx)])
+            i += 1
+
+        # Binary search for 3 percentiles
+        q1, med, q2 = map(lambda val: bisect(lambda x: probf(x) - val, left, right, maxiter=10000, xtol=0.00001),
+                            [0.25, 0.5, 0.75])
```

Comment:
this is a bit hard to read. How about using a function
```suggestion
        def search(val: float) -> float:
            return bisect(lambda x: probf(x) - val, left, right, maxiter=10000, xtol=0.00001)

        q1, med, q2 = map(search, [0.25, 0.5, 0.75])
```